src/cryptsetup/cryptsetup-generator.c

```diff
@@ -795,10 +795,56 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
         return 0;
 }
 
+static int add_crypttab_device(const char *name, const char *device,  const char *keyspec, const char *options) {
+        _cleanup_free_ char *keyfile = NULL, *keydev = NULL, *headerdev = NULL, *filtered_header = NULL;
+        crypto_device *d = NULL;
+        char *uuid;
+        int r;
+
+        uuid = startswith(device, "UUID=");
+        if (!uuid)
+                uuid = path_startswith(device, "/dev/disk/by-uuid/");
+        if (!uuid)
+                uuid = startswith(name, "luks-");
+        if (uuid)
+                d = hashmap_get(arg_disks, uuid);
+
+        if (arg_allow_list && !d) {
+                log_info("Not creating device '%s' because it was not specified on the kernel command line.", name);
+                return 0;
+        }
+
+        r = split_locationspec(keyspec, &keyfile, &keydev);
+        if (r < 0)
+                return r;
+
+        if (options && (!d || !d->options)) {
+                r = filter_header_device(options, &headerdev, &filtered_header);
+                if (r < 0)
+                        return r;
+                options = filtered_header;
+        }
+
+        r = create_disk(name,
+                        device,
+                        keyfile,
+                        keydev,
+                        (d && d->options) ? d->headerdev : headerdev,
```

Comment:
Not your fault, but the check should be `d && d->headerdev`?

Reply:
I don't think so. IIUC the intent is to prefer options of an existing device and `headerdev` is extracted from `options`. IOW, "if `d->options` is set, use `d->options` and `d->headerdev`, otherwise use the local values".

Reply:
At least, this does not change the existing behavior. So, this PR itself is OK.